### PR TITLE
Remove nonexisting entries from classpath

### LINKED ISSUE #11587: Gradle should not add non-existing paths to class paths
If you unzip [CDI-test.zip](https://github.com/gradle/gradle/files/3926493/CDI-test.zip) and execute the test task, the CDI container initialization fails, as there are non-existing paths in the class path.

If you open `build.gradle.kts` and comment-in the commented-out lines that create any non-existing paths, the test runs through without problem.

But what is the sense of adding non-existing paths to any class path. I think Gradle should prevent adding non-existing paths to class paths, not only when running tests, but also when starting other Java tasks to prevent this and similar problems.

### PR DESCRIPTION
**Pull request by towith**
Fixes #11587 

### Context

https://github.com/gradle/gradle/issues/11587#issue-533280628

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#git-commit---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation
- [x] Recognize contributor in release notes

### CODE REVIEW

**Comment by JLLeitschuh** [comment]
Hi @towith,
Would you be willing contribute a simple test to verify this logic. Also, do you have a small reproducer that demonstrates the bug you are fixing?

**Comment by towith** [comment]
@JLLeitschuh  ok, i have commit a integration test class `org.gradle.process.internal.worker.DefaultWorkerProcessBuilderTest` to repo, with the  change the test will pass , and without the change  it fails , the integration test is base on issue #11587

**Comment by JLLeitschuh** [comment]
@bot-gradle test this please

**Comment by JLLeitschuh** [comment]
@towith, this looks like it has introduced test failures. Please explore.

**Comment by towith** [comment]
> @towith, this looks like it has introduced test failures. Please explore.



> @towith, this looks like it has introduced test failures. Please explore.

Here are messages cause the failure:
```
succeeds("test")
|
org.gradle.integtests.fixtures.executer.UnexpectedBuildFailure: org.gradle.internal.exceptions.LocationAwareException: Execution failed for task ':compileTestGroovy'.
	at org.gradle.integtests.fixtures.executer.InProcessGradleExecuter.doRun(InProcessGradleExecuter.java:161)
Caused by: org.gradle.internal.exceptions.LocationAwareException: Execution failed for task ':compileTestGroovy'.
Caused by: java.lang.NoClassDefFoundError: Could not initialize class org.codehaus.groovy.reflection.ReflectionCache
	java.lang.NoClassDefFoundError: Could not initialize class org.codehaus.groovy.vmplugin.v7.Java7

* What went wrong:
Execution failed for task ':compileTestGroovy'.
> Could not initialize class org.codehaus.groovy.reflection.ReflectionCache
```

seems cause by  groovy failed to compile the script 
I find a similar issue at https://github.com/gradle/gradle/issues/8681

@JLLeitschuh I'll study more about this failure, thanks for response

**Comment by towith** [comment]
@JLLeitschuh by the way , i can find the jdk version is 1.8, but i can not find out version of groovy , what version of groovy lib in the failed test is ?

**Comment by towith** [comment]
> @towith, this looks like it has introduced test failures. Please explore.

Hi, @JLLeitschuh  I have explored these failure logs , there should be two tests failed ,  one is  `JUnitIntegrationTest.executesTestsInCorrectEnvironment` , i have change the code of test logic  so that file path in testClasspath should exists, and this test can pass now:
image

The other test failure is `org.gradle.process.internal.worker.DefaultWorkerProcessBuilderTest.test classpath exists` , which i can i got it passed on my machine, 
image

I had noticed the test failed in bot use `jdk version 15` , I have try to installed  one , but run the test got the message ,

image

So  I cannot reproduce the failed second test on my machine as on test bots, could you provide some instruction about how should i do  to make the test failure happened  as on test bot machine? Thanks

**Comment by JLLeitschuh** [comment]
@gradle/core I have a feeling that if we want to fix this, it would be better to fix this at a higher level than this? Maybe I'm missing something? It looks like other API's around this method have a similar issue (eg, `applicationModulePath`)?

Also, are there any cases where `applicationClasspath` would get a file passed to it that doesn't exist yet, but will exist when the classpath is used?

**Comment by towith** [comment]
@JLLeitschuh  Maybe your concern does make sense, I agree with you.

**Comment by Vampire** [comment]
Any news here? :-)

**Comment by Vampire** [comment]
@towith @JLLeitschuh and news with this?

**Comment by towith** [comment]
@Vampire Sorry, I have not check github notifications for long time, just see what you post  now.  

Cause of this issue is some "target application" gradle called check path in classpath, and if not exists it throws exception.

Unfortunately,  The pull request code as solution is to check every path in classpath, if not exists, then not add to the classpath,  But as  jjohannes  said, this may restrict the flexibility of gradle, what if some "target application" need non-exist path in classpath for specific purpose, such as dynamic load class at runtime, So this pull request maybe not the fine solution. For this reason, maybe your work round you have posted is better.
```
    tasks.test {
                doFirst {
                   classpath
                          .filterNot { it.exists() }
                           .forEach { it.mkdirs() }
               }
            }
```

If you agree, I would close this issue, Thinks for your concern!

**Comment by Vampire** [comment]
I know the cause of the issue, the issue you are fixing here is reported by me. ;-)

I personally don't really agree, that's why I opened that issue.
And I also don't see Jonathan saying this is an issue, just thinking about whether this is an issue and on the other hand even thinking about doing an even wider approach.

Actually the definition of the class path says:
> Classpath entries that are neither directories nor archives (.zip or .jar files) nor * are ignored.

So it is of course also a bug in the piece that throws exception like the CDI container I used.
But there is also not much point in adding anything that does not adhere to those criteria which non-existing paths don't do. :-)

**Comment by towith** [comment]
@Vampire Appreciate your rigorous  attitude towards technical things,   But personally  I prefer gradle should remain flexibility , Because I think the build tool  should  be of more application concern , hence should cover as wider  option for user as possible. That's why I think your work round is better solution later. It should works and not add restrictions upon user's choice, that is by the work round, user can decide the classpath contains non-exist path or not.

**Comment by Vampire** [comment]
Don't get me wrong, I'm always in favor of flexibility.
But not so much for pointless flexibility.
The definition of class path says that other entries are ignored, so there is no point in adding those other entries.

If the Gradle guys think it should stay like this, I'm fine with living with this.
But it is just pointless adding things to the class path that are going to be (or should be if there was no bug in Weld) ignored then anyway.
This just makes the class path unnecessary longer which can actually lead to problems, especially on Windows with its very limit default max commandline length.

**Comment by towith** [comment]
Well , we both have expressed  what we thought about this , And it's up to gradle  members then.  Nice to talk with you, probably  now is at night in Germany as in China, Good night, Vampire :)

File: subprojects/core/src/integTest/groovy/org/gradle/process/internal/worker/DefaultWorkerProcessBuilderTest.groovy [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+ * limitations under the License.
+ */
+
+package org.gradle.process.internal.worker
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+
+class DefaultWorkerProcessBuilderTest extends AbstractIntegrationSpec {
+
+    def "test classpath exists"() {
```
Comment (donat) [inline subprojects/core/src/integTest/groovy/org/gradle/process/internal/worker/DefaultWorkerProcessBuilderTest.groovy:23]: ❌ 

```suggestion
    def "test classpath does not contain nonexistent entries"() {
```

File: subprojects/core/src/integTest/groovy/org/gradle/process/internal/worker/DefaultWorkerProcessBuilderTest.groovy [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+                    given:
+                        def seContainer = SeContainerInitializer.newInstance()
+                                .addProperty('javax.enterprise.inject.scan.implicit', TRUE)
+                                .initialize()
+
+                    expect:
+                        seContainer.select(Main).get().dependencies.size() == 1
+
+                    cleanup:
+                        seContainer?.close()
```
Comment (donat) [inline subprojects/core/src/integTest/groovy/org/gradle/process/internal/worker/DefaultWorkerProcessBuilderTest.groovy:76]: ❌ Although the CDI example exhibits the problem, we can spare quite some CI resources by simplifying the test. You can just write an assertion that all items in `System.getProperty('java.class.path')` exists.

File: subprojects/core/src/integTest/groovy/org/gradle/process/internal/worker/DefaultWorkerProcessBuilderTest.groovy [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            dependencies {
+                implementation("javax.enterprise:cdi-api:2.0")
+
+                testImplementation("org.spockframework:spock-core:1.3-groovy-2.5")
+
+                testRuntimeOnly("org.jboss.weld.se:weld-se-core:3.1.3.Final")
+                testRuntimeOnly("org.jboss:jandex:2.1.1.Final")
+            }
+
+            //tasks.test {
```
Comment (donat) [inline subprojects/core/src/integTest/groovy/org/gradle/process/internal/worker/DefaultWorkerProcessBuilderTest.groovy:102]: ❌ remove debug code

File: subprojects/core/src/integTest/groovy/org/gradle/process/internal/worker/DefaultWorkerProcessBuilderTest.groovy [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+                @ApplicationScoped
+                static class TestDependency implements Dependency {
+                }
+            }
+        '''
+
+        buildFile << '''
+            apply plugin: "java"
+            apply plugin: "groovy"
```
Comment (donat) [inline subprojects/core/src/integTest/groovy/org/gradle/process/internal/worker/DefaultWorkerProcessBuilderTest.groovy:87]:
```suggestion
            plugins {
                id "java-library"
                id "groovy"
            }
```

**Review by donat — CHANGES_REQUESTED** [review]

File: subprojects/core/src/main/java/org/gradle/process/internal/worker/DefaultWorkerProcessBuilder.java [hunk 5]
```diff
@@ -100,7 +100,14 @@ public String getBaseName() {
 
     @Override
     public WorkerProcessBuilder applicationClasspath(Iterable<File> files) {
-        GUtil.addToCollection(applicationClasspath, files);
+        for (File file : files) {
+            if (file == null) {
+                throw new IllegalArgumentException("Illegal null value provided in this collection: " + files);
+            }
+            if (file.exists()) {
```
Comment (Vampire) [inline subprojects/core/src/main/java/org/gradle/process/internal/worker/DefaultWorkerProcessBuilder.java:107]: If this is also the code path used by things like `JavaExec` and similar, it should also allow wildcards, and then should probably be something along the lines of `file.exists() || file.getName().equals("*")`.

Comment (towith) [reply subprojects/core/src/main/java/org/gradle/process/internal/worker/DefaultWorkerProcessBuilder.java:107]: Thinks, I will take some spare time rewrite test and the if statement,  What does the `*` wildcards stands for, does it mean for a specific path?

Comment (Vampire) [reply subprojects/core/src/main/java/org/gradle/process/internal/worker/DefaultWorkerProcessBuilder.java:107]: `foo/bar/*` on a class path is effectively the same as listing all jars in that directory explicitly. On case-sensitive filesystems like per default on Windows, it should - consider all `jar` and `JAR` if I remember correctly.

Comment (towith) [reply subprojects/core/src/main/java/org/gradle/process/internal/worker/DefaultWorkerProcessBuilder.java:107]: Sorry misunderstanding  this  as whole path as `*` , Yes, It should be considered.

**Comment by towith** [comment]
@donat  I have rewrite the code, especially tests, please check and run test again to see if it works

**Comment by donat** [comment]
@bot-gradle test rFM

**Comment by donat** [comment]
@bot-gradle test rFM

**Comment by donat** [comment]
@bot-gradle test rFM

**Comment by donat** [comment]
@bot-gradle test rFM

**Comment by donat** [comment]
@bot-gradle test and merge

**Review by donat — APPROVED** [review]

**Comment by donat** [comment]
@towith thank you for your contribution!

**Comment by towith** [comment]
@donat you're welcome : )

**Pull request merged**